### DISCUSSION TITLE
User Activity table

### DISCUSSION THREAD

**Discussion by omaroudat** [riclolsen/json-scada#55] [Q&A]
Hi Mr. ricardo, 

I have a Question please: 

I managed to export the User Actions table with all the details, but I need to know the to have the current login role for the user who did the action 

something like GetCurrentRole()

Appreciate your support

**Comment by riclolsen** [comment]
Hi Omar,

Sorry, I couldn't fully understand your question.
Users and roles are connected by the users.roles array of ObjectIds of the roles collection.
There is no information in the tables about the users currently logged in the system. When a user logs in the systems a JWT token is given to its browser. As long as this token is valid, requests from that user's browser are considered valid. So there is no session established, no concept of logged in users.
You may mean the roles a user had when doing a logged action. In this case there is no easy answer. User's roles can change and the roles themselves can also change. All those changes are recorded as user actions also when done via UI. You have to review those possible changes to find what the roles the user had and what were the roles' rights when the action was issued. Anyway, the action was permitted to the user, so I do not know if this is relevant.

**Reply by omaroudat** [reply]
Hi Ricardo, 

Thank you for your answer. 

  What I mean, as per the attached screenshot below, the user action contains username, Action, Tag Name, etc. 

some of the users may have different roles for example Omar have "Admin" Role and User "Role" 
He did an action to Ack the Alarm, What I want to achieve is to capture the role that the system used for him to do this action

So I want to capture per each event what is the role used for this user to perform the action 


![image](https://user-images.githubusercontent.com/27445284/149622702-c0acd302-83be-4b2a-9036-22e837bd1258.png)


![image](https://user-images.githubusercontent.com/27445284/149622835-ff9b90d7-e68b-4f84-95bf-cb6dbf3e0a69.png)

**Reply by riclolsen** [reply]
I see.
The user has a set of roles.
Each role has a set of rights.
The rights of a user are defined by the most relaxed rights of its roles.
Each action is authorized or not depending on user's rights.
It is not registered the roles or rights a user had when the action is issued.

User's roles can change and the roles themselves can also change. All those changes are recorded as user actions also when done via UI. You have to review those possible changes to find what the roles the user had and what were the roles' rights when the action was issued. 

You can change the code here (line 1031) near de end of this file

     src/server_realtime_auth/app/controllers/auth.controller.js

to record also the roles the user has
    
    // enqueue use action for later insertion to mongodb
    function registerUserAction (req, actionName) {
      let body = {}
      Object.assign(body, req.body)
    
      let ck = checkToken(req)
      if (ck !== false) {
        console.log(actionName + ' - ' + ck?.username)
        delete body['password']
        // register user action
        UserActionsQueue.enqueue({
          username: ck?.username,
          roles: ck?.roles,                 // add this line
          properties: body,
          action: actionName,
          timeTag: new Date()
        })
      } else {
        console.log(actionName + ' - ' + req.body?.username)
        delete body['password']
        // register user action
        UserActionsQueue.enqueue({
          username: req.body?.username,
          properties: body,
          action: actionName,
          timeTag: new Date()
        })
      }
    }

**Comment by omaroudat** [comment]
Thank you dear for your reply. 

I added the line to the auth.controller.js file. 

but the results in MongoDB tables still the same even for the new actions as shown below: 
![image](https://user-images.githubusercontent.com/27445284/150643832-44257007-fa87-471b-9840-11a64845805a.png)


![image](https://user-images.githubusercontent.com/27445284/150643865-3dbe7ed9-19fd-4bff-abfc-39b79bb6146f.png)


Anything missing needs to be added  ?

**Reply by riclolsen** [reply]
Try to restart the server_realtime_auth service.

**Reply by omaroudat** [reply]
I already tried that even I restarted the Server. But still the same results 

The Collection is json_scada.userActions

**Reply by omaroudat** [reply]
Hi @riclolsen 

I tried to restart the service multiple times and I restarted the server as well but with no luck. 

Appreciate if you can help

**Reply by riclolsen** [reply]
Sorry, this is really not enough.
Replace 2 files attached here.
C:\json-scada\src\server_realtime_auth\app\controllers\auth.controller.js
C:\json-scada\src\server_realtime_auth\index.js
Rename the files to javascript extension ".js" .
Restart the server_realtime_auth service.
Logout and login the user to recreate JWT tokens.
Then it should work.


[index.js.txt](https://github.com/riclolsen/json-scada/files/7958388/index.js.txt)
[auth.controller.js.txt](https://github.com/riclolsen/json-scada/files/7958393/auth.controller.js.txt)
.
